File: backend/app/schemas/cliente.py
"""
Schemas de Cliente.
"""

import re
from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List
from pydantic import BaseModel, Field, EmailStr, field_validator


CUIT_PATTERN = r"^\d{2}-\d{8}-\d{1}$"


def _validar_cuit_opcional(v: Optional[str]) -> Optional[str]:
    """Valida formato XX-XXXXXXXX-X. Vacío/None se acepta (campo opcional)."""
    if not v:
        return v
    v = v.strip()
    if not v:
        return None
    if not re.match(CUIT_PATTERN, v):
        raise ValueError("CUIT debe tener formato XX-XXXXXXXX-X")
    return v


# ==================== CLIENTE ====================

class ClienteBase(BaseModel):
    """Schema base de cliente. Los datos fiscales viven en TitularFiscal."""
    tipo: str = "particular"
    razon_social: str = Field(..., min_length=2, max_length=200)
    nombre_fantasia: Optional[str] = None
    titular_fiscal_id: Optional[str] = None
    # cuit/condicion_iva son "atajos": el service crea o reutiliza el
    # TitularFiscal correspondiente. Un mismo CUIT puede tener varios
    # clientes (grupo empresarial), se agrupan bajo el mismo titular.
    cuit: Optional[str] = None
    condicion_iva: Optional[str] = None
    email: Optional[EmailStr] = None
    telefono: Optional[str] = None
    celular: Optional[str] = None
    contacto_nombre: Optional[str] = None
    contacto_cargo: Optional[str] = None
    direccion: Optional[str] = None
    ciudad: Optional[str] = None
    provincia: str = "Córdoba"
    codigo_postal: Optional[str] = None
    descuento_general: Optional[Decimal] = Field(default=0, ge=0, le=100)
    limite_credito: Optional[Decimal] = None
    dias_credito: Optional[int] = Field(default=0, ge=0)
    dia_retiro_preferido: Optional[str] = None
    horario_retiro_preferido: Optional[str] = None
    requiere_factura: bool = False
    enviar_notificaciones: bool = True
    notas: Optional[str] = None
    notas_internas: Optional[str] = None

    @field_validator("cuit")
    @classmethod
    def _cuit_format(cls, v: Optional[str]) -> Optional[str]:
        return _validar_cuit_opcional(v)


class ClienteCreate(ClienteBase):
    """Schema para crear cliente."""
    pass


class ClienteUpdate(BaseModel):
    """Schema para actualizar cliente."""
    tipo: Optional[str] = None
    razon_social: Optional[str] = Field(None, min_length=2, max_length=200)
    nombre_fantasia: Optional[str] = None
    titular_fiscal_id: Optional[str] = None
    # Ver ClienteBase.cuit: atajo que resuelve/crea el TitularFiscal en el service.
    cuit: Optional[str] = None
    condicion_iva: Optional[str] = None
    email: Optional[EmailStr] = None
    telefono: Optional[str] = None
    celular: Optional[str] = None
    contacto_nombre: Optional[str] = None
    contacto_cargo: Optional[str] = None
    direccion: Optional[str] = None
    ciudad: Optional[str] = None
    provincia: Optional[str] = None
    codigo_postal: Optional[str] = None
    lista_precios_id: Optional[str] = None
    descuento_general: Optional[Decimal] = Field(default=None, ge=0, le=100)
    limite_credito: Optional[Decimal] = None
    dias_credito: Optional[int] = Field(default=None, ge=0)
    dia_retiro_preferido: Optional[str] = None
    horario_retiro_preferido: Optional[str] = None
    requiere_factura: Optional[bool] = None
    enviar_notificaciones: Optional[bool] = None
    notas: Optional[str] = None
    notas_internas: Optional[str] = None
    activo: Optional[bool] = None

    @field_validator("cuit")
    @classmethod
    def _cuit_format(cls, v: Optional[str]) -> Optional[str]:
        return _validar_cuit_opcional(v)


class ClienteResponse(ClienteBase):
    """Schema de respuesta de cliente."""
    id: str
    codigo: str
    lista_precios_id: Optional[str] = None
    saldo_cuenta_corriente: Decimal
    fecha_alta: Optional[date] = None
    fecha_ultima_compra: Optional[date] = None
    activo: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    # Datos fiscales delegados al titular (properties del modelo)
    cuit: Optional[str] = None
    condicion_iva: str = "consumidor_final"

    # Calculados
    nombre_display: str
    tiene_deuda: bool
    supera_limite_credito: bool

    class Config:
        from_attributes = True


class ClienteList(BaseModel):
    """Schema simplificado para listados."""
    id: str
    codigo: str
    tipo: str
    razon_social: str
    nombre_fantasia: Optional[str] = None
    cuit: Optional[str] = None  # delegado al titular vía property
    email: Optional[str] = None
    telefono: Optional[str] = None
    ciudad: Optional[str] = None
    saldo_cuenta_corriente: Decimal
    activo: bool
    tiene_deuda: bool
    titular_fiscal_id: Optional[str] = None

    class Config:
        from_attributes = True


class ClienteSelect(BaseModel):
    """Schema para selectores/dropdowns."""
    id: str
    codigo: str
    nombre: str  # nombre_display
    cuit: Optional[str] = None  # delegado al titular vía property
    lista_precios_id: Optional[str] = None

    class Config:
        from_attributes = True


# ==================== TIPOS Y CONSTANTES ====================

TIPOS_CLIENTE = [
    {"value": "particular", "label": "Particular"},
    {"value": "empresa", "label": "Empresa"},
    {"value": "hotel", "label": "Hotel"},
    {"value": "restaurante", "label": "Restaurante"},
    {"value": "hospital", "label": "Hospital"},
    {"value": "gimnasio", "label": "Gimnasio"},
    {"value": "otro", "label": "Otro"},
]

CONDICIONES_IVA = [
    {"value": "responsable_inscripto", "label": "Responsable Inscripto"},
    {"value": "monotributo", "label": "Monotributo"},
    {"value": "exento", "label": "Exento"},
    {"value": "consumidor_final", "label": "Consumidor Final"},
    {"value": "no_responsable", "label": "No Responsable"},
]

File: frontend/src/types/cliente.ts
/**
 * Tipos para el módulo de Clientes
 */

// Estados y enums
export type TipoCliente = 'particular' | 'empresa' | 'hotel' | 'restaurante' | 'hospital' | 'gimnasio' | 'otro';
export type CondicionIVA = 'responsable_inscripto' | 'monotributo' | 'exento' | 'consumidor_final' | 'no_responsable';
export type EstadoPedido = 'borrador' | 'confirmado' | 'en_proceso' | 'listo' | 'entregado' | 'facturado' | 'cancelado';
export type TipoEntrega = 'retiro_local' | 'delivery' | 'envio';
export type TipoMovimientoCC = 'cargo' | 'pago' | 'ajuste';
export type MedioPago = 'efectivo' | 'transferencia' | 'tarjeta_debito' | 'tarjeta_credito' | 'cheque' | 'mercado_pago' | 'cuenta_corriente' | 'otro';

// Cliente
export interface Cliente {
  id: string;
  codigo: string;
  tipo: TipoCliente;
  razon_social: string;
  nombre_fantasia: string | null;
  cuit: string | null;
  condicion_iva: CondicionIVA;
  email: string | null;
  telefono: string | null;
  celular: string | null;
  contacto_nombre: string | null;
  contacto_cargo: string | null;
  direccion: string | null;
  ciudad: string | null;
  provincia: string;
  codigo_postal: string | null;
  lista_precios_id: string | null;
  descuento_general: number | null;
  limite_credito: number | null;
  dias_credito: number | null;
  saldo_cuenta_corriente: number;
  dia_retiro_preferido: string | null;
  horario_retiro_preferido: string | null;
  requiere_factura: boolean;
  enviar_notificaciones: boolean;
  fecha_alta: string | null;
  fecha_ultima_compra: string | null;
  notas: string | null;
  notas_internas: string | null;
  activo: boolean;
  created_at: string;
  updated_at: string | null;
  // Calculados
  nombre_display: string;
  tiene_deuda: boolean;
  supera_limite_credito: boolean;
}

export interface ClienteList {
  id: string;
  codigo: string;
  tipo: TipoCliente;
  razon_social: string;
  nombre_fantasia: string | null;
  cuit: string | null;
  email: string | null;
  telefono: string | null;
  ciudad: string | null;
  saldo_cuenta_corriente: number;
  activo: boolean;
  tiene_deuda: boolean;
}

export interface ClienteCreate {
  tipo?: TipoCliente;
  razon_social: string;
  nombre_fantasia?: string | null;
  cuit?: string | null;
  condicion_iva?: CondicionIVA;
  email?: string | null;
  telefono?: string | null;
  celular?: string | null;
  contacto_nombre?: string | null;
  contacto_cargo?: string | null;
  direccion?: string | null;
  ciudad?: string | null;
  provincia?: string;
  codigo_postal?: string | null;
  descuento_general?: number | null;
  limite_credito?: number | null;
  dias_credito?: number | null;
  dia_retiro_preferido?: string | null;
  horario_retiro_preferido?: string | null;
  requiere_factura?: boolean;
  enviar_notificaciones?: boolean;
  notas?: string | null;
  notas_internas?: string | null;
}

// Pedido
export interface DetallePedido {
  id: string;
  servicio_id: string | null;
  descripcion: string;
  cantidad: number;
  unidad: string;
  precio_unitario: number;
  descuento_porcentaje: number | null;
  subtotal: number;
  notas: string | null;
}

export interface Pedido {
  id: string;
  numero: string;
  cliente_id: string;
  estado: EstadoPedido;
  fecha_pedido: string;
  fecha_retiro: string | null;
  fecha_entrega_estimada: string | null;
  fecha_entrega_real: string | null;
  fecha_facturacion: string | null;
  tipo_entrega: TipoEntrega;
  direccion_entrega: string | null;
  horario_entrega: string | null;
  subtotal: number;
  descuento_porcentaje: number | null;
  descuento_monto: number | null;
  iva: number;
  total: number;
  saldo_pendiente: number;
  factura_numero: string | null;
  factura_tipo: string | null;
  notas: string | null;
  notas_internas: string | null;
  observaciones_entrega: string | null;
  creado_por_id: string;
  created_at: string;
  updated_at: string | null;
  detalles: DetallePedido[];
  cliente_nombre: string | null;
  creado_por_nombre: string | null;
}

export interface PedidoList {
  id: string;
  numero: string;
  cliente_id: string;
  cliente_nombre: string | null;
  estado: EstadoPedido;
  fecha_pedido: string;
  fecha_retiro: string | null;
  fecha_entrega_estimada: string | null;
  total: number;
  saldo_pendiente: number;
  tipo_entrega: TipoEntrega;
}

export interface DetallePedidoCreate {
  servicio_id?: string | null;
  descripcion: string;
  cantidad: number;
  unidad?: string;
  precio_unitario: number;
  descuento_porcentaje?: number | null;
  notas?: string | null;
}

export interface PedidoCreate {
  cliente_id: string;
  fecha_pedido: string;
  fecha_retiro?: string | null;
  fecha_entrega_estimada?: string | null;
  tipo_entrega?: TipoEntrega;
  direccion_entrega?: string | null;
  horario_entrega?: string | null;
  descuento_porcentaje?: number | null;
  notas?: string | null;
  notas_internas?: string | null;
  observaciones_entrega?: string | null;
  detalles?: DetallePedidoCreate[];
}

// Cuenta Corriente
export interface MovimientoCuentaCorriente {
  id: string;
  tipo: TipoMovimientoCC;
  concepto: string;
  monto: number;
  fecha_movimiento: string;
  saldo_anterior: number;
  saldo_posterior: number;
  factura_numero: string | null;
  recibo_numero: string | null;
  medio_pago: MedioPago | null;
  referencia_pago: string | null;
}

export interface EstadoCuenta {
  cliente_id: string;
  cliente_nombre: string;
  saldo_actual: number;
  limite_credito: number | null;
  credito_disponible: number | null;
  total_facturado_mes: number;
  total_pagado_mes: number;
  cantidad_facturas_pendientes: number;
  factura_mas_antigua_dias: number | null;
}

export interface RegistrarPagoRequest {
  monto: number;
  fecha: string;
  medio_pago: MedioPago;
  referencia_pago?: string | null;
  notas?: string | null;
  aplicar_a_pedidos?: string[];
}

// Constantes
export const TIPOS_CLIENTE = [
  { value: 'particular', label: 'Particular' },
  { value: 'empresa', label: 'Empresa' },
  { value: 'hotel', label: 'Hotel' },
  { value: 'restaurante', label: 'Restaurante' },
  { value: 'hospital', label: 'Hospital' },
  { value: 'gimnasio', label: 'Gimnasio' },
  { value: 'otro', label: 'Otro' },
];

export const CONDICIONES_IVA = [
  { value: 'responsable_inscripto', label: 'Responsable Inscripto' },
  { value: 'monotributo', label: 'Monotributo' },
  { value: 'exento', label: 'Exento' },
  { value: 'consumidor_final', label: 'Consumidor Final' },
  { value: 'no_responsable', label: 'No Responsable' },
];

export const ESTADOS_PEDIDO = [
  { value: 'borrador', label: 'Borrador', color: 'gray' },
  { value: 'confirmado', label: 'Confirmado', color: 'blue' },
  { value: 'en_proceso', label: 'En Proceso', color: 'yellow' },
  { value: 'listo', label: 'Listo', color: 'green' },
  { value: 'entregado', label: 'Entregado', color: 'purple' },
  { value: 'facturado', label: 'Facturado', color: 'teal' },
  { value: 'cancelado', label: 'Cancelado', color: 'red' },
];

export const TIPOS_ENTREGA = [
  { value: 'retiro_local', label: 'Retiro en Local' },
  { value: 'delivery', label: 'Delivery' },
  { value: 'envio', label: 'Envío' },
];

export const MEDIOS_PAGO = [
  { value: 'efectivo', label: 'Efectivo' },
  { value: 'transferencia', label: 'Transferencia Bancaria' },
  { value: 'tarjeta_debito', label: 'Tarjeta de Débito' },
  { value: 'tarjeta_credito', label: 'Tarjeta de Crédito' },
  { value: 'cheque', label: 'Cheque' },
  { value: 'mercado_pago', label: 'Mercado Pago' },
  { value: 'cuenta_corriente', label: 'Cuenta Corriente' },
  { value: 'otro', label: 'Otro' },
];
